### ISSUE TITLE
CA2227:CollectionPropertiesShouldBeReadOnly is raised on immutable collection properties

### ISSUE THREAD

**Issue by airbreather** [dotnet/roslyn-analyzers#1900]
## Analyzer package
[Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

## Package Version

v2.6.2

## Diagnostic ID

[CA2227:CollectionPropertiesShouldBeReadOnly](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca2227-collection-properties-should-be-read-only)

## Repro steps
Create a new project with the following:

### Sample.cs
```csharp
using System.Collections.Immutable;

public sealed class Sample
{
    public string Name { get; set; }

    // Line in question:
    public ImmutableArray<byte> ImmutableData { get; set; }
}
```

### ConsoleApp0.csproj
```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>netcoreapp2.1</TargetFramework>
    <LangVersion>7.2</LangVersion>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.CodeAnalysis.FxCopAnalyzers" Version="2.6.2" />
  </ItemGroup>

</Project>
```

## Expected behavior
I feel like members with types from `System.Collections.Immutable`, such as the one on the "line in question", should not raise `CA2227:CollectionPropertiesShouldBeReadOnly`.

I agree with raising it for `System.Collections.ObjectModel.ReadOnlyCollection<T>`, but the contract of `System.Collections.Immutable` types is a promise that once you have an instance of an immutable collection, **nobody** should be able to edit its contents.

## Actual behavior
The "line in question" raises `CA2227:CollectionPropertiesShouldBeReadOnly`.

**Comment by mavasani** [comment]
@airbreather Note that the violation is being fired for the property itself `ImmutableData`, not the contents of the immutable data it stores, i.e. a public client can still assign a new immutable array to `c.ImmutableData = ...`, which is undesirable. The recommendation is to remove the setter for the property or make it non-public. Closing this as by design.. Let us know if you still have any confusion.

**Comment by airbreather** [comment]
> Note that the violation is being fired for the property itself `ImmutableData`, not the contents of the immutable data it stores, i.e. a public client can still assign a new immutable array to `c.ImmutableData = ...`, which is undesirable.

@mavasani What makes `c.ImmutableData = ...` more undesirable than `c.Name = ...`?

**Comment by mavasani** [comment]
The [rule description](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca2227-collection-properties-should-be-read-only?view=vs-2017#rule-description) and comments in the [example](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca2227-collection-properties-should-be-read-only?view=vs-2017#example) clarifies the purpose of this rule, which is to avoid exposing _writable collections_. Feel to suppress or disable this rule in your ruleset if you feel it is fine to expose writable collections for your source code.

**Comment by mavasani** [comment]
@airbreather Ah thanks for the detailed explanation. I did misinterpret your request, and this does seem like something we can special case for this rule.
